test/test.identitypoolclient.ts

```diff
@@ -0,0 +1,530 @@
+import {
+  BaseExternalAccountClient,
+  IamGenerateAccessTokenResponse,
+} from '../src/auth/baseexternalclient';
+
+nock.disableNetConnect();
+
+interface IamGenerateAccessTokenError {
+  error: {
+    code: number;
+    message: string;
+    status: string;
+  };
+}
+
+interface NockMockStsToken {
+  statusCode: number;
+  response: StsSuccessfulResponse | OAuthErrorResponse;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  request: {[key: string]: any};
+  additionalHeaders?: {[key: string]: string};
+}
+
+interface NockMockGenerateAccessToken {
+  statusCode: number;
+  token: string;
+  response: IamGenerateAccessTokenResponse | IamGenerateAccessTokenError;
+  scopes: string[];
+}
+
+const ONE_HOUR_IN_SECS = 3600;
+const baseUrl = 'https://sts.googleapis.com';
+const path = '/v1/token';
+const saEmail = 'service-1234@service-name.iam.gserviceaccount.com';
+const saBaseUrl = 'https://iamcredentials.googleapis.com';
+const saPath = `/v1/projects/-/serviceAccounts/${saEmail}:generateAccessToken`;
+
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Escaping
+function escapeRegExp(str: string): string {
+  // $& means the whole matched string.
+  return str.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&');
+}
+
+function mockStsTokenExchange(nockParams: NockMockStsToken[]): nock.Scope {
+  const scope = nock(baseUrl);
+  nockParams.forEach(nockMockStsToken => {
+    const headers = Object.assign(
+      {
+        'content-type': 'application/x-www-form-urlencoded',
+      },
+      nockMockStsToken.additionalHeaders || {}
+    );
+    scope
+      .post(path, qs.stringify(nockMockStsToken.request), {
+        reqheaders: headers,
+      })
+      .reply(nockMockStsToken.statusCode, nockMockStsToken.response);
+  });
+  return scope;
+}
+
+function mockGenerateAccessToken(
+  nockParams: NockMockGenerateAccessToken[]
+): nock.Scope {
+  const scope = nock(saBaseUrl);
+  nockParams.forEach(nockMockGenerateAccessToken => {
+    const token = nockMockGenerateAccessToken.token;
+    scope
+      .post(
+        saPath,
+        {
+          scope: nockMockGenerateAccessToken.scopes,
+        },
+        {
+          reqheaders: {
+            Authorization: `Bearer ${token}`,
+            'Content-Type': 'application/json',
+          },
+        }
+      )
+      .reply(
+        nockMockGenerateAccessToken.statusCode,
+        nockMockGenerateAccessToken.response
+      );
+  });
+  return scope;
+}
+
+function assertGaxiosResponsePresent(resp: GetAccessTokenResponse) {
+  const gaxiosResponse = resp.res || {};
+  assert('data' in gaxiosResponse && 'status' in gaxiosResponse);
+}
+
+describe('IdentityPoolClient', () => {
+  const fileSubjectToken = fs.readFileSync(
+    './test/fixtures/external-subject-token.txt',
+    'utf-8'
+  );
+  const projectNumber = '123456';
+  const poolId = 'POOL_ID';
+  const providerId = 'PROVIDER_ID';
+  const audience =
```

Comment:
If all these variables are used by the `audience` URL, and not elsewhere in test, I would be tempted to encapsulate them in a test helper, like `getAudienceUrl`; I read this and my immediate reaction was, that's a lot of variables, before I realized they're not all used in test cases.

Reply:
Ok, I made some refactor and moved a lot of these constants to common file and exported helper functions which are used by 2 test files.